File: src/agent/tools/builtin/read-config.ts
import type { ToolHandler } from "../../../types";
import { getConfig } from "../../../config/index";

/**
 * Read the current openagent.json configuration (with secrets masked).
 */
export const readConfigTool: ToolHandler = {
  definition: {
    name: "read_config",
    description:
      "Read the current openagent.json configuration. " +
      "API keys and secrets are masked for safety. " +
      "Use this to check current settings before making changes with write_config.",
    parameters: {
      type: "object",
      properties: {},
      required: [],
    },
  },

  async execute() {
    const config = getConfig();

    // Deep clone and mask secrets
    const safe = JSON.parse(JSON.stringify(config)) as Record<string, unknown>;
    const providers = safe.providers as Record<string, Record<string, unknown>> | undefined;
    if (providers) {
      for (const p of Object.values(providers)) {
        if (typeof p.apiKey === "string" && p.apiKey) {
          p.apiKey = p.apiKey.slice(0, 8) + "...";
        }
        if (typeof p.setupToken === "string" && p.setupToken) {
          p.setupToken = p.setupToken.slice(0, 8) + "...";
        }
      }
    }
    const channels = safe.channels as Record<string, Record<string, unknown>> | undefined;
    if (channels) {
      for (const ch of Object.values(channels)) {
        if (typeof ch.appSecret === "string" && ch.appSecret) {
          ch.appSecret = ch.appSecret.slice(0, 4) + "...";
        }
        if (typeof ch.botToken === "string" && ch.botToken) {
          ch.botToken = ch.botToken.slice(0, 8) + "...";
        }
      }
    }

    return JSON.stringify(safe, null, 2);
  },
};

File: src/tools/builtins/read_config.py
"""Read config tool, matching TS tools/builtin/read-config.ts."""

from __future__ import annotations

import json

from src.types import ToolDefinition, ToolHandler


def _mask(value: str, keep: int = 8) -> str:
    if not value or len(value) <= keep:
        return value
    return value[:keep] + "..."


async def _execute(args: dict) -> str:
    from src.config import get_config

    config = get_config()
    raw = config.model_dump(by_alias=True)

    # Mask sensitive fields
    if p := raw.get("providers"):
        if oai := p.get("openai"):
            if oai.get("apiKey"):
                oai["apiKey"] = _mask(oai["apiKey"])
        if ant := p.get("anthropic"):
            if ant.get("apiKey"):
                ant["apiKey"] = _mask(ant["apiKey"])
            if ant.get("setupToken"):
                ant["setupToken"] = _mask(ant["setupToken"])
    if ch := raw.get("channels"):
        if feishu := ch.get("feishu"):
            if feishu.get("appSecret"):
                feishu["appSecret"] = _mask(feishu["appSecret"], 4)
        if tg := ch.get("telegram"):
            if tg.get("botToken"):
                tg["botToken"] = _mask(tg["botToken"])

    return json.dumps(raw, indent=2)


read_config_tool = ToolHandler(
    definition=ToolDefinition(
        name="read_config",
        description="Read the current OpenAgent configuration (API keys are masked).",
        parameters={"type": "object", "properties": {}},
    ),
    execute=_execute,
)
